Superplot - small changes

gui-co opened the pull request:
**Description of work.**

This PR introduces few requested changes in the Superplot widget:
* new context menu entry for superplot with error bars
* fixed size for the add/remove button. This will avoid any change in button width when the text is modified
* remove the warning saying that `getname()` is deprecated
* when remove is pressed, the current position of the slider is kept

This also solves the two issues mentioned in #31968

**To test:**

 Check the new features.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
[SPLOT] getName() is deprecated

Commit:
[SPLOT] keep selection when hold is released

Commit:
[SPLOT] fix hold button size

Because the button text is changing, the size was also changing. This
commit fixes the button size from the largest text.

Commit:
[SPLOT] keep superplot button active

When the fit browser can not be started, the superplot button must
remain active.

Commit:
[SPLOT] error bar is a superplot parameter

Commit:
[SPLOT] separate function to enable error bars

Commit:
[SPLOT] enable error bars in superplot

Commit:
[SPLOT] superplot with errors in context menu

Commit:
[SPLOT] connect superplot with errors menu

Commit:
[SPLOT] superplot in context menu with bin plot

Commit:
[SPLOT] plot function too complex (flake8)

gvardany commented:
This mostly works, just one issue is that if a workspace is a single bin, the superplot should start with `Bin` as the choice in the combobox. For the Fit and Superplot buttons, now when one is clicked, the other is disabled; one has to unclick the other one first. Is there a way to directly switch with one click?

Commit:
[SPLOT] context menu entries for superplot bins

Commit:
[SPLOT] connect context menu entries for bin mode

Commit:
[SPLOT] setter for bin/spectrum mode

Commit:
[SPLOT] fit and superplot mutually exclusive

With this commit, there is no need to close the fit browser or the
superplot before opening the other. The buttons are linked together and
opening one will automatically close the other.

Commit:
[SPLOT] naming consistency in context menu

Commit:
[SPLOT] update slider with bins number

Commit:
[SPLOT] adapt tooltip to button text

gvardany reviewed (dismissed):
All to spec now, great work. Good to go once the builds pass.

Commit:
[SPLOT] button size has to be ajusted to be OK

gui-co commented:
There was an issue with the add/remove button size. It is fixed now.

Commit:
[SPLOT] dimension names match users specifications

DavidFair commented on Framework/PythonInterface/mantid/plots/plotfunctions.py:
```diff
@@ -217,8 +218,10 @@ def plot(workspaces, spectrum_nums=None, wksp_indices=None, errors=False,
-            workspace_names = [ws.getName() for ws in workspaces]
+            workspace_names = [ws.name() for ws in workspaces]
             fig.canvas.manager.superplot.set_workspaces(workspace_names)
+            fig.canvas.manager.superplot.set_bin_mode(plot_kwargs and "axis" in plot_kwargs and plot_kwargs["axis"] == MantidAxType.BIN)
```
Can we add a test to `plotfunctionTest.py` to ensure `set_bin_mode` is called correctly for both the case where `MantixAxType.BIN` is passed in the nested dict, and that it gracefully handles cases where it we don't.

Currently, this looks like it handles all edge cases. But because of the complexity of accessing a nested item if someone does some overzealous refactoring they might break this subtly in the future without us noticing easily.

gui-co replied:
I added a test.

DavidFair commented on qt/applications/workbench/workbench/plotting/figuremanager.py:
```diff
@@ -399,10 +399,11 @@ def fit_toggle(self):
+            self.toolbar._actions["toggle_fit"].setChecked(False)
         else:
+            if self.toolbar._actions["toggle_superplot"].isChecked():
+                self.superplot_toggle()
```
This is me bike shedding, but toggle implies switching state on/off, should this be called `superplot_enable()` instead?
The current method name looks like we should call toggle every time a user toggles the toolbar button on / off currently, rather than just leaving it on

gui-co replied:
I split the toggle function in two: `superplot_show` and `superplot_hide`; and called `superplot_hide` here.

DavidFair commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -53,6 +58,30 @@ def __init__(self, canvas, parent=None):
         self._update_spectrum_slider()
         self._update_hold_button()
 
+    def set_bin_mode(self, state):
+        """
+        Set the plot mode.
+
+        Args:
+            state (bool): if true, bin mode is set, if false, spectrum mode
+        """
+        if state:
+            self._view.set_mode(self.BIN_MODE_TEXT)
+        else:
+            self._view.set_mode(self.SPECTRUM_MODE_TEXT)
+        self._update_spectrum_slider()
+        self._update_plot()
+
+    def enable_error_bars(self, state):
+        """
+        Enable/disable error bars in plot.
+
+        Args:
+            state (bool): if True, the error bars will be on
+        """
+        self._error_bars = state
+        self._update_plot()
```
These new methods are missing tests. 
We should especially check `self._update_plot` is called (just substitute the method with a mock within the test) as that's very easy to remove, introducing visual bugs that are only found with manual testing.

gui-co replied:
I added the missing tests.

DavidFair commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -53,6 +58,30 @@ def __init__(self, canvas, parent=None):
+        if state:
+            self._view.set_mode(self.BIN_MODE_TEXT)
+        else:
+            self._view.set_mode(self.SPECTRUM_MODE_TEXT)
```
This if needs to move to the model as per our [GUI standards](https://developer.mantidproject.org/Standards/GUIStandards.html#no-matter-how-small-logic-goes-into-the-model), and have a test associated.

It's a simple if statement, so it's a pain, but we agreed to do this for all logic as it tends to grow 

Something like this would be fine:
```
self._view.set_mode(self.model.getSpectrumMode(state))
```

gui-co replied:
I am not sure to understand which logic should go in the model. Here, this is only a setter for the state of the view when the superplot is started (i.e. the selected item of the combobox).

DavidFair replied:
Sorry with fresh eyes I've just seen its `BIN_MODE_TEXT` rather than `BIN_MODE` which I originally read it as.
Feel free to ignore this, as I assumed it was switching modes rather than text displays

DavidFair reviewed (changes requested):
Some minor project standards issues.

Overall it looks good and the improved wording on user facing elements improves the readability significantly

Commit:
[SPLOT] missing tests

Commit:
[SPLOT] add test for superplot on bin plot

Commit:
Merge branch 'master' into superplot_corrections

Commit:
[SPLOT] split toggle function in "hide" and "show"

DavidFair reviewed (commented):
Thanks for doing those changes :+1:, `superplot_show` and `superplot_hide` clarify the operations significantly.

I was just re-testing what this fixes but it appears this change isn't working for me:
> when remove is pressed, the current position of the slider is kept

To replicate:
- Create two sample workspaces
- Open super plot with errors on just ws1
- Add ws2 using the add button in the top left
- Drag the slider to an arbitrary point
- Remove either ws1 or ws2 using the remove icon in superplot
- This will reset the displayed plot to a grey screen and the current position of the slider is lost

Is this a regression or a known problem @gui-co ?

Commit:
[SPLOT] slider update only when del selected item

The deletion of a workspace triggers the update of the slider only if
the deleted worspace was the selected workspace.

Commit:
[SPLOT] bug - update when deletion is selection

gui-co commented:
You are right. The position of the slider was kept only when a spectrum was removed (Add/Remove button) and not when the workspace was deleted.

It should be fixed now.

Commit:
[SPLOT] add test for the deletion of selection

DavidFair reviewed (approved):
LGTM and resolved the bug I was seeing, thanks :+1:
